# Feature Engineering for Medical Insurance Cost Prediction

In [1]:
import pandas as pd
df=pd.read_csv("../data/insurance.csv")

## Load the Dataset

The first step in Feature Engineering is to load the original dataset. We use the raw dataset so that every preprocessing step is applied consistently and independently. This ensures that the notebook is reproducible and does not depend on the output of previous notebooks.

In [2]:
df.head()

,age,sex,bmi,children,smoker,region,charges
0,19,female,27.900,0,yes,southwest,16884.92400
1,18,male,33.770,1,no,southeast,1725.55230
2,28,male,33.000,3,no,southeast,4449.46200
3,33,male,22.705,0,no,northwest,21984.47061
4,32,male,28.880,0,no,northwest,3866.85520


### Observation

- The dataset has been loaded successfully.
- It contains both input features and the target variable (`charges`).
- The raw dataset will now be transformed into a format suitable for machine learning through feature engineering.

In [8]:
X=df.drop("charges",axis=1)
y=df["charges"]


,age,sex,bmi,children,smoker,region
0,19,female,27.900,0,yes,southwest
1,18,male,33.770,1,no,southeast
2,28,male,33.000,3,no,southeast
3,33,male,22.705,0,no,northwest
4,32,male,28.880,0,no,northwest
...,...,...,...,...,...,...
1333,50,male,30.970,3,no,northwest
1334,18,female,31.920,0,no,northeast
1335,18,female,36.850,0,no,southeast
1336,21,female,25.800,0,no,southwest


### Observation

- The input features have been stored in `X`.
- The target variable (`charges`) has been stored in `y`.
- The model will use `X` to learn patterns and predict `y`.

In [13]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42)
y_train.shape

(1070,)

In [14]:
print("X_train:", X_train.shape)
print("X_test :", X_test.shape)
print("y_train:", y_train.shape)
print("y_test :", y_test.shape)

X_train: (1070, 6)
X_test : (268, 6)
y_train: (1070,)
y_test : (268,)


### Observation

- The dataset has been successfully divided into training and testing sets.
- 80% of the data (1070 samples) will be used to train the model.
- 20% of the data (268 samples) will be used to evaluate the model on unseen data.
- Splitting the data before preprocessing helps prevent data leakage and provides a fair evaluation of the model.

In [15]:
X_train

,age,sex,bmi,children,smoker,region
560,46,female,19.950,2,no,northwest
1285,47,female,24.320,0,no,northeast
1142,52,female,24.860,0,no,southeast
969,39,female,34.320,5,no,southeast
486,54,female,21.470,3,no,northwest
...,...,...,...,...,...,...
1095,18,female,31.350,4,no,northeast
1130,39,female,23.870,5,no,southeast
1294,58,male,25.175,0,no,northeast
860,37,female,47.600,2,yes,southwest


In [19]:
categorical_features=["sex","region","smoker"]
print("categorical_features:",categorical_features)
numerical_features =["age","bmi","children"]
print("numerical_features:",numerical_features)

categorical_features: ['sex', 'region', 'smoker']
numerical_features: ['age', 'bmi', 'children']


### Observation

- The dataset contains **3 categorical features** (`sex`, `smoker`, and `region`) that require encoding before model training.
- It also contains **3 numerical features** (`age`, `bmi`, and `children`) that will later be scaled.
- Identifying feature types helps us apply the correct preprocessing technique to each column.

Binary Encoding

We do Encoding the sex and smoker in the binary form because they have only two category and Linear Regression needs the numerical data so we have to encoding but we choose binary encoding due to less feature value and also we can't drop this columns beacuse they are the important feautures for charges which is dependent on this faetures 

In [32]:
X_train["sex"]=X_train["sex"].map({"female":1,"male":0})
X_train["smoker"]=X_train["smoker"].map({"yes":1,"no":0})
X_test["sex"]=X_test["sex"].map({"female":1,"male":0})
X_test["smoker"]=X_test["smoker"].map({"yes":1,"no":0})
X_train

,age,sex,bmi,children,smoker,region
560,46,NaN,19.950,2,NaN,northwest
1285,47,NaN,24.320,0,NaN,northeast
1142,52,NaN,24.860,0,NaN,southeast
969,39,NaN,34.320,5,NaN,southeast
486,54,NaN,21.470,3,NaN,northwest
...,...,...,...,...,...,...
1095,18,NaN,31.350,4,NaN,northeast
1130,39,NaN,23.870,5,NaN,southeast
1294,58,NaN,25.175,0,NaN,northeast
860,37,NaN,47.600,2,NaN,southwest


Observations
In sex the following represntation is follow

1.male-0

2.feamle-1

In smoker

1.yes-1

2.no-0
Binary encoding converted the categorical features sex and smoker into numerical values so that they can be used by machine learning algorithms. Since both features have only two categories, binary encoding is sufficient and avoids creating unnecessary additional columns.

In [33]:
X_train=pd.get_dummies(X_train,columns=["region"],drop_first=True,dtype=int)
X_test=pd.get_dummies(X_test,columns=["region"],drop_first=True,dtype=int)
X_train

,age,sex,bmi,children,smoker,region_northwest,region_southeast,region_southwest
560,46,NaN,19.950,2,NaN,1,0,0
1285,47,NaN,24.320,0,NaN,0,0,0
1142,52,NaN,24.860,0,NaN,0,1,0
969,39,NaN,34.320,5,NaN,0,1,0
486,54,NaN,21.470,3,NaN,1,0,0
...,...,...,...,...,...,...,...,...
1095,18,NaN,31.350,4,NaN,0,0,0
1130,39,NaN,23.870,5,NaN,0,1,0
1294,58,NaN,25.175,0,NaN,0,0,0
860,37,NaN,47.600,2,NaN,0,0,1


Observations 
During One-Hot Encoding, one category can always be inferred from the remaining encoded columns. Keeping all dummy variables introduces perfect multicollinearity, where one feature is a linear combination of the others. This can make regression coefficients unstable and difficult to interpret. Therefore, we drop one dummy column while preserving all the information.
So we only have three coulmns instead of 4

## Feature Scaling

Feature scaling is one of the most important part of the feature engineering because they make the model performace better  .The model has the fixed range not too large difference of value so which made it easy and relevent 

,age,sex,bmi,children,smoker,region,charges
0,19,female,27.900,0,yes,southwest,16884.92400
1,18,male,33.770,1,no,southeast,1725.55230
2,28,male,33.000,3,no,southeast,4449.46200
3,33,male,22.705,0,no,northwest,21984.47061
4,32,male,28.880,0,no,northwest,3866.85520
...,...,...,...,...,...,...,...
1333,50,male,30.970,3,no,northwest,10600.54830
1334,18,female,31.920,0,no,northeast,2205.98080
1335,18,female,36.850,0,no,southeast,1629.83350
1336,21,female,25.800,0,no,southwest,2007.94500
